src/project.cc

```diff
@@ -470,94 +470,99 @@ void Project::Load(const std::string &root) {
-    if (it != folder.path2entry_index.end()) {
-      Project::Entry &entry = folder.entries[it->second];
-      if (can_redirect || entry.filename == path)
-        return entry;
-      if (entry.compdb_size) {
-        best_folder = &folder;
-        best = &entry;
-      }
-      break;
-    }
-  }
+  Project::Folder *best_compdb_folder = nullptr;
 
-  bool exists = false;
-  std::string dir;
-  const std::vector<const char *> *extra = nullptr;
   Project::Entry ret;
+  std::lock_guard lock(mtx);
+
   for (auto &[root, folder] : root2folder)
-    if (StringRef(path).startswith(root))
+    if (StringRef(path).startswith(root)) {
+      // Find the best-fit .ccls
       for (auto &[dir1, args] : folder.dot_ccls)
-        if (StringRef(path).startswith(dir1)) {
-          dir = dir1;
-          extra = &args;
-          if (AppendToCDB(args))
-            goto out;
-          exists = true;
-
-          ret.root = ret.directory = root;
-          ret.filename = path;
-          if (args.empty()) {
-            ret.args = GetFallback(path);
-          } else {
-            ret.args = args;
-            ret.args.push_back(Intern(path));
-          }
-          ProjectProcessor(folder).Process(ret);
-          for (const std::string &arg : g_config->clang.extraArgs)
-            ret.args.push_back(Intern(arg));
-          ret.args.push_back(Intern("-working-directory=" + ret.directory));
-          return ret;
+        if (StringRef(path).startswith(dir1) &&
+            dir1.length() > best_dot_ccls_dir.length()) {
+          best_dot_ccls_root = root;
+          best_dot_ccls_folder = &folder;
+          best_dot_ccls_dir = dir1;
+          best_dot_ccls_args = &args;
         }
-out:
-  if (must_exist && !exists)
-    return ret;
 
-  if (!best) {
-    int best_score = INT_MIN;
-    for (auto &[root, folder] : root2folder) {
-      if (dir.size() && !StringRef(path).startswith(dir))
-        continue;
-      for (const Entry &e : folder.entries)
-        if (e.compdb_size) {
-          int score = ComputeGuessScore(path, e.filename);
-          if (score > best_score) {
-            best_score = score;
-            best = &e;
-            best_folder = &folder;
+      if (!exact_match) {
+        auto it = folder.path2entry_index.find(path);
+        if (it != folder.path2entry_index.end()) {
+          Project::Entry &entry = folder.entries[it->second];
+          exact_match = can_redirect || entry.filename == path;
+          if (exact_match || entry.compdb_size) {
+            best_compdb_folder = &folder;
+            best = &entry;
           }
         }
+      }
     }
-  }
 
-  ret.is_inferred = true;
-  ret.filename = path;
-  if (!best) {
-    if (ret.root.empty())
-      ret.root = g_config->fallbackFolder;
-    ret.directory = ret.root;
-    ret.args = GetFallback(path);
+  bool append;
+  if (best_dot_ccls_args && !(append = AppendToCDB(*best_dot_ccls_args))) {
+    // Override the compdb entry.
```

Comment:
It is stated on [wiki](https://github.com/MaskRay/ccls/wiki/Project-Setup#compile_commandsjson-1) that 
> By default .ccls compiler flags are applied only to files not listed in compile_commands.json.

However this is going to override the compile db entry.

Reply:
This rule may be confusing. I think it is probably clearer to let `.ccls` override `compile_commands.json` if the first option is not `%compile_commands.json`